### PR TITLE
feat: add support for multiple trees per file

### LINKED ISSUE #50: feature request: add support for different targeting behaviours
Currently, Bulloak works with `.tree` files that target a single specific function. This is great for a small number of contracts or a subset of functionality to be tested but doesn't scale when there are lots of contracts and functions.

Ideally, it should be possible to scaffold on a per-contract basis, targeting multiple functions in a single `.tree` file that is emitted to the same test contract. Consider the following ERC20 tree as an example:

```tree
ERC20.t.sol
├── when decimals is called
│   └── it should return the decimals
├── when name is called
│   └── it should return the name
├── when symbol is called
│   └── it should return the symbol
├── when totalSupply is called
│   └── it should return the total supply
├── when balanceOf is called
│   └── it should return the balance of the address
├── when allowance is called
│   └── it should return the allowance of the owner to the spender
├── when mint is called
│   ├── when the caller does not have the minter role
│   │   └── it should revert
│   └── when the caller has the minter role
│       ├── given the address is the zero address
│       │   └── it should revert
│       └── given the address is not the zero address
│           ├── given the amount causes the total supply to overflow
│           │   └── it should revert
│           └── given the amount does not cause the total supply to overflow
│               ├── it should mint the tokens
│               ├── it should update the total supply
│               └── it should emit a {Transfer} event
├── when burn is called
│   ├── when the caller does not have sufficient balance
│   │   └── it should revert
│   └── when the caller has sufficient balance
│       ├── it should burn the tokens
│       ├── it should update the total supply
│       └── it should emit a {Transfer} event
├── when transfer is called
│   ├── given the address is the zero address
│   │   └── it should revert
│   └── given the address is not the zero address
│       ├── when the sender does not have sufficient balance
│       │   └── it should revert
│       └── when the sender has sufficient balance
│           ├── it should transfer the tokens
│           ├── it should emit a {Transfer} event
│           └── it should return true
├── when approve is called
│   ├── given the address is the zero address
│   │   └── it should revert
│   └── given the address is not the zero address
│       ├── it should set the allowance
│       ├── it should emit an {Approval} event
│       └── it should return true
├── when transferFrom is called
│   ├── given the from address is the zero address
│   │   └── it should revert
│   └── given the from address is not the zero address
│       ├── given the to address is the zero address
│       │   └── it should revert
│       └── given the to address is not the zero address
│           ├── given the amount is zero
│           │   └── it should revert
│           └── given the amount is not zero
│               └── when the spender is not the from address
│                   ├── when the spender has maximum allowance
│                   │   ├── it should transfer the tokens
│                   │   ├── it should emit a {Transfer} event
│                   │   └── it should return true
│                   └── when the spender does not have maximum allowance
│                       ├── when the spender does not have sufficient allowance
│                       │   └── it should revert
│                       └── when the spender has sufficient allowance
│                           ├── it should decrement the allowance
│                           ├── it should transfer the tokens
│                           ├── it should emit a {Transfer} event
│                           └── it should return true
```

For maximum flexibility of the directory structure, it may also be useful to specify the trees to be emitted to the same test contract. For example, it may be desirable to have all `.tree` files within some parent `trees` directory, further organized into subdirectories (e.g., `token/ERC20`, `libraries/Strings, `vendor/AccessControl`), which can be arbitrarily combined to scaffold any number of test files.

### PR DESCRIPTION
**Pull request by giovannidisiena**
WIP attempt at handling multiple trees per `.tree` file. Currently, it appears I have broken some of the existing tests but I'm not yet sure how/why.

Steps taken:
- [x] Find all roots in a .tree by adding a new `TokenKind::Root`.
- [x] Parse & translate each AST individually.
- [x] Combiner: Check that all roots belong to the same contract, i.e. Contract for each root Contract.function is the same.
- [x] Combiner: Merge all the HIRs by collecting the function nodes and appending them to the root contract node.
- [x] Combine and report any errors.
- [x] Add tests for all of the above.
- [x] Add roots check to bulloak check - this is done by `hir::translate` when `check::context::Context` is created since check mostly works with parse trees derived from HIRs.

@alexfertel it would be great to get some initial feedback here, when you get some time please, to make sure I'm not completely barking up the wrong tree or making too much of a mess. I have left a few `follow-up` comments for both my and your benefit - I'm sure there's lots here that could be improved or otherwise written in a more idiomatic fashion so please do let me know where that is the case as I'm definitely still getting to grips with Rust!

Partially closes #50.

### CODE REVIEW

**Comment by alexfertel** [comment]
This is looking very promising! Thanks for such an effort.

Reading the code made me realize that we might be able to do something simpler: What if we parallelized the inner phases? That is, we currently roughly have the following compiler pipeline:

```
.tree -> &str -> scaffold entrypoint -> tokenize & parse -> semantic analysis -> HIR -> PT -> emit 
```

The above works for one (1) tree. Would you be willing to try feeding each separate tree to the pipeline? You would only touch two points in the pipeline:

```
.tree -> &str -> split input -> scaffold entrypoint -> ... -> HIR -> Combine -> PT -> emit
                 ^^^^^^^^^^^                                         ^^^^^^^
```
Which ends up looking like:
```
                                scaffold entrypoint -> ... -> Hir 
                              /                                   \
.tree -> &str -> split input -> scaffold entrypoint -> ... -> Hir -> Combine -> PT -> emit
                              \                                   /
                                scaffold entrypoint -> ... -> Hir 
```


Where the split is a simple:

```rust
let trees = tree.split("\n\n").collect::<Vec<_>>();
``` 

And the Combine phase looks similar to what you have.

If this makes sense to you I think it would simplify a lot. Wdyt?

On the other hand, I think the combine step is a good place for verifying that all the trees have the same contract name 👌🏻

File: src/hir/combiner.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+        Ok(())
+    }
+
+    /// Combines the translated HIRs into a single HIR.
+    ///
+    /// This function is called after the ASTs are translated to HIR.
+    pub fn combine(&self, hirs: Vec<Hir>) -> Hir {
+        // Merge all the HIRs by collecting the function nodes and appending them to the root contract node.
+        // @follow-up - how do we deal with duplicate modifiers in separate HIRs?
```
Comment (alexfertel) [inline src/hir/combiner.rs:52]: We move it to the combine phase, which is this one.

File: src/hir/mod.rs [hunk 2]
```diff
@@ -12,8 +13,22 @@ pub use hir::*;
...
 pub fn translate(tree: &str) -> anyhow::Result<Hir> {
-    let ast = crate::syntax::parse(tree)?;
+    let asts = crate::syntax::parse(tree)?;
+    let combiner = combiner::Combiner::new();
+    match combiner.verify(&asts) {
+        Err(_e) => {
+            // @follow-up - propagate this error
+        }
+        _ => {}
+    }
```
Comment (alexfertel) [inline src/hir/mod.rs:23]:
```suggestion
    combiner.verify(&asts)?;
```

Another option is:

```suggestion
    if let Err(e) = combiner.verify(&asts) {
 		// Do something with err
    }
```

You'll need to use `anyhow::Result` instead of `Result<(), String>` in line 23 of `src/hir/combiner.rs` if you want to propagate it.

File: src/hir/mod.rs [hunk 3]
```diff
@@ -12,8 +13,22 @@ pub use hir::*;
...
+    let mut hirs = Vec::new();
     let mut discoverer = crate::scaffold::modifiers::ModifierDiscoverer::new();
-    let modifiers = discoverer.discover(&ast);
-    Ok(translator::Translator::new().translate(&ast, modifiers))
+    let translator = translator::Translator::new();
+    for ast in &asts {
+        // @follow-up - use `map` instead of `for _ in`, here and elsewhere?
```
Comment (alexfertel) [inline src/hir/mod.rs:28]: It's fine to use for loops. Iterators are usually faster, but `bulloak` doesn't need to be super performant and a for loop is more readable oftentimes.

File: src/hir/mod.rs [hunk 4]
```diff
@@ -12,8 +13,22 @@ pub use hir::*;
...
+        let modifiers = discoverer.discover(ast).unwrap();
+        hirs.push(translator.clone().translate(ast, modifiers));
```
Comment (alexfertel) [inline src/hir/mod.rs:30]:
```suggestion
        hirs.push(translator.translate(ast, modifiers));
```

Let's try to avoid cloning here by changing https://github.com/alexfertel/bulloak/blob/c38e7795c6a018019ed2e2e0712868b5db0202be/src/hir/translator.rs#L30 to accept `&self` instead of `self`.

**Review by alexfertel — COMMENTED** [review]
A few small comments answering `@follow-up`.

**Comment by giovannidisiena** [comment]
Thanks @alexfertel, the review was very helpful - I have just pushed a new commit with the simplified pipeline. I have also identified the cause of most of the existing test failures due to some slight modifications I made to the use of identifier mode in `syntax::tokenizer`, which I believe to be correct but have downstream effects, so if you could take a look at the remaining updated `follow-up` comments and provide your thoughts please that would be awesome.

File: src/hir/mod.rs [hunk 5]
```diff
@@ -11,9 +12,32 @@ pub use hir::*;
 ///
 /// This function leverages `crate::syntax::parse` and `translator::Translator::translate`
 /// to hide away most of the complexity of `bulloak`'s internal compiler.
-pub fn translate(tree: &str) -> anyhow::Result<Hir> {
-    let ast = crate::syntax::parse(tree)?;
+pub fn translate(text: &str) -> anyhow::Result<Hir> {
```
Comment (alexfertel) [inline src/hir/mod.rs:15]: So the idea is that this function doesn't have to change at all.

```rust
fn entrypoint() -> anyhow::Result<Hir> {
	// This is the &str read from the file.
	let trees = "TREES";
	let hirs = trees.split(TREE_SEPARATOR).map(|t| translate(t)?);
	Combiner::new().combine(&hirs)	
}

fn translate(text: &str) -> anyhow::Result<Hir> {
	...
}
```

The above is illustrative, so it needs some tweaks, but that is the overall idea.

Comment (alexfertel) [reply src/hir/mod.rs:15]: I think we can start by just reporting the first error, and we can handle collecting all the errors later so that you don't have to deal with it right now.

File: src/scaffold/emitter.rs [hunk 6]
```diff
@@ -244,18 +244,41 @@ mod tests {
...
+        let hirs: Vec<Hir> = asts
+            .iter()
+            .map(|ast| {
+                let modifiers = discoverer.discover(ast);
+                translator.translate(ast, modifiers)
+            })
+            .collect();
+
+        combiner.verify(&asts).unwrap();
+        let hir = combiner.combine(&hirs);
```
Comment (alexfertel) [inline src/scaffold/emitter.rs:281]: So, for this one it would be the same as before, but the result would be different. We would create a second function that calls this one for each tree, collects the hirs, and emits. This means that this one would be almost identical to `translate` + `entrypoint` as above, but adding the emitter step.

File: src/scaffold/mod.rs [hunk 7]
```diff
@@ -110,10 +110,28 @@ impl<'s> Scaffolder<'s> {
 
     /// Generates Solidity code from a `.tree` file.
     pub fn scaffold(&self, text: &str) -> crate::error::Result<String> {
-        let ast = syntax::parse(text)?;
+        let combiner = hir::combiner::Combiner::new();
```
Comment (alexfertel) [inline src/scaffold/mod.rs:113]: Same here, changes should be minimal: we split this function into more functions so that the pipeline described above is clearer to the reader.

File: src/scaffold/mod.rs [hunk 8]
```diff
@@ -110,10 +110,28 @@ impl<'s> Scaffolder<'s> {
...
+
         let mut discoverer = modifiers::ModifierDiscoverer::new();
-        let modifiers = discoverer.discover(&ast);
-        let hir = hir::translator::Translator::new().translate(&ast, modifiers);
+        let translator = hir::translator::Translator::new();
+
+        let mut errors = Vec::new();
+        let asts: Vec<syntax::ast::Ast> = trees
+            .iter()
+            .filter_map(|&tree| syntax::parse(tree).map_err(|error| errors.push(error)).ok()) // @follow-up - how do we combine and report errors?
```
Comment (alexfertel) [inline src/scaffold/mod.rs:122]: You can skip this for now and just report the first one with `?`.

File: src/scaffold/mod.rs [hunk 9]
```diff
@@ -110,10 +110,28 @@ impl<'s> Scaffolder<'s> {
...
+
+        let hirs: Vec<hir::Hir> = asts
+            .iter()
+            .map(|ast| {
+                let modifiers = discoverer.discover(ast);
+                translator.translate(ast, modifiers)
+            })
+            .collect();
+
+        combiner.verify(&asts).unwrap();
```
Comment (alexfertel) [inline src/scaffold/mod.rs:133]: We need to be very careful with unwraps, because they are very frustrating for users. At the very least, this should be an `expect`. However, in this particular case, we should forward the error to the callee so that it decides how to handle it.

File: src/syntax/tokenizer.rs [hunk 10]
```diff
@@ -329,7 +331,9 @@ impl<'s, T: Borrow<Tokenizer>> TokenizerI<'s, T> {
                 }
                 _ => {
                     let token = self.scan_word()?;
-                    if token.kind == TokenKind::When || token.kind == TokenKind::Given {
+                    if token.kind == TokenKind::When || token.kind == TokenKind::Given
+                    // || token.kind == TokenKind::It // @follow-up - I think this is needed, but it breaks the tests because of invalid {} character in emit event statement
```
Comment (alexfertel) [inline src/syntax/tokenizer.rs:335]: This is not correct. Actions can contain any character because when emitted they are just comments.

Could you expand on why it's needed?

File: src/syntax/tokenizer.rs [hunk 11]
```diff
@@ -140,7 +140,8 @@ impl Tokenizer {
         Self {
             pos: Cell::new(Position::new(0, 1, 1)),
             // Starts as `true` because the first token must always be a contract name.
-            identifier_mode: Cell::new(true),
+            // identifier_mode: Cell::new(true), // @follow-up - this is correct, but it breaks the tests because of invalid . character in root
```
Comment (alexfertel) [inline src/syntax/tokenizer.rs:143]: The `.`s in a root should be valid now, since I am calling `sanitize` on contract names. We have multiple tests for this, but the main one that covers this functionality you can find here:

https://github.com/alexfertel/bulloak/blob/c38e7795c6a018019ed2e2e0712868b5db0202be/tests/scaffold.rs#L12

Which, scaffolds this file:

https://github.com/alexfertel/bulloak/blob/c38e7795c6a018019ed2e2e0712868b5db0202be/tests/scaffold/basic.tree

Which contains a contract name with a `.`.

File: tests/check.rs [hunk 12]
```diff
@@ -117,6 +117,8 @@ fn checks_invalid_tree() {
     let output = cmd(&binary_path, "check", &tree_path, &[]);
     let actual = String::from_utf8(output.stderr).unwrap();
 
+    println!("{}", actual); // the tree is missing its matching Solidity file
```
Comment (alexfertel) [inline tests/check.rs:120]: 👀

File: tests/scaffold.rs [hunk 13]
```diff
@@ -58,6 +58,7 @@ fn errors_when_tree_is_empty() {
         let output = cmd(&binary_path, "scaffold", &tree_path, &[]);
         let actual = String::from_utf8(output.stderr).unwrap();
 
+        println!("{}", actual);
```
Comment (alexfertel) [inline tests/scaffold.rs:61]: 👀

**Review by alexfertel — CHANGES_REQUESTED** [review]
Heyo! This is indeed a step forward, thanks for taking the time.

I gave this one a closer look since I think that we can commit to the design. Lmk if you have any questions!

File: src/hir/combiner.rs [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+                            _ => root.children.push(child.clone()),
+                        }
+                    }
+                }
+                _ => unreachable!(),
+            }
+        }
+
+        Hir::Root(root)
+    }
```
[the remark this answers is not in the document]
Comment (giovannidisiena) [reply src/hir/combiner.rs:89]: No advantage, I think I just originally split these steps when breaking down the problem into its constituent parts but agree that they should be performed together in a single function.

File: src/syntax/tokenizer.rs (hunk 10 above)
Comment (giovannidisiena) [reply src/syntax/tokenizer.rs:335]: Ah, yes. Ignore this, apologies - I originally thought identifier mode was required for scanning all `Word` tokens but then equally saw that other characters in the actions would then be considered invalid, so this makes sense given that they are emitted as comments.

Comment (alexfertel) [reply src/syntax/tokenizer.rs:335]: No need to apologize at all!

File: src/scaffold/mod.rs (hunk 7 above)
Comment (giovannidisiena) [reply src/scaffold/mod.rs:113]: Sorry, so to confirm, here `Scaffolder::scaffold` is the main entrypoint which is being called via `Scaffold::run` - are you suggesting that we add a new `entrypoint` function (similar to the above) that is called instead? Something like this:
```
fn entrypoint() -> crate::error::Result<String> {
    // This is the &str read from the file.
    let trees = "TREES";
    let hirs = trees.split(TREE_SEPARATOR).map(|t| self.scaffold(t)?);
    Combiner::new().combine(&hirs);
    let pt = sol::Translator::new(self.solidity_version).translate(&hir);
    let source = sol::Formatter::new().emit(pt);
    let formatted = fmt(&source).expect("should format the emitted solidity code");
    Ok(formatted)
}

pub fn scaffold(&self, text: &str) -> anyhow::Result<Hir> {
    let ast = syntax::parse(text)?;
    let mut discoverer = modifiers::ModifierDiscoverer::new();
    let modifiers = discoverer.discover(&ast);
    let hir = hir::translator::Translator::new().translate(&ast, modifiers);
    Ok(hir)
}
```

Comment (alexfertel) [reply src/scaffold/mod.rs:113]: I should have been more clear. Something like:

```rust
    pub fn scaffold(&self, text: &str) -> crate::error::Result<String> {
        let hirs = text.split(TREE_SEPARATOR).map(|t| self.get_hir(t)?);
        Combiner::new().combine(&hirs);
        let pt = sol::Translator::new(self.solidity_version).translate(&hir);
        let source = sol::Formatter::new().emit(pt);
        let formatted = fmt(&source).expect("should format the emitted solidity code");

        Ok(formatted)
    }

    fn get_hir(&self, text: &str) -> anyhow::Result<hir::Hir> {
        let ast = syntax::parse(text)?;
        let mut discoverer = modifiers::ModifierDiscoverer::new();
        let modifiers = discoverer.discover(&ast);
        let hir = hir::translator::Translator::new().translate(&ast, modifiers);
        Ok(hir)
    }
```

Comment (alexfertel) [reply src/scaffold/mod.rs:113]: `get_hir` maybe isn't the best name, but wanted to give a better quickly

Comment (alexfertel) [reply src/scaffold/mod.rs:113]: As you can tell, `get_hir` is basically the same as `translate` in `src/hir/mod.rs`.

Comment (giovannidisiena) [reply src/scaffold/mod.rs:113]: Understood, thanks 👍

File: src/hir/combiner.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (giovannidisiena) [reply src/hir/combiner.rs:89]: Ah, actually, having looked at this again more closely, I split these steps i. so that we fail early if the contract names do not match, and ii. because it was easier to verify the `contract_name` member on `Ast::Root` rather than finding all the `Hir::ContractDefinition`s and comparing `identifier`s. In doing so, I just realized that we actually don't want to push all children that are `Hir::ContractDefinition`s to the `Root` but rather only one of these and the rest of the `Hir::FunctionDefinition`s pushed to this child, so I need to update this anyway and will modify the verification to work with identifiers instead.

[the remark this answers is not in the document]
Comment (alexfertel) [reply src/hir/combiner.rs:89]: > I just realized that we actually don't want to push all children that are Hir::ContractDefinitions to the Root but rather only one of these and the rest of the Hir::FunctionDefinitions pushed to this child, so I need to update this anyway and will modify the verification to work with identifiers instead.

This makes sense to me 👍🏻 

> i. so that we fail early if the contract names do not match

Just want to comment here that I think we don't want to fail early here: As a user, I'd like to see all the contract name mismatches at once, instead of one at a time. It can be frustrating to fix the name in one tree only to run bulloak again and find out that there was another name mismatch.

This doesn't need to be implemented here, but it's just something to keep in mind.

[the remark this answers is not in the document]
Comment (giovannidisiena) [reply src/hir/combiner.rs:89]: Great! And I meant fail early in the sense that we verify all contract names match before combining then but yeah realise that the current implementation throws mismatch errors one at a time so will fix that with the other error-handling suggestions elsewhere.

File: src/hir/combiner.rs [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+            for ast in asts {
+                if let Ast::Root(root) = ast {
+                    let parts: Vec<&str> =
+                        root.contract_name.split(|c| c == '.' || c == ':').collect();
+                    if parts[0] != first_parts[0] {
+                        return Err(format!(
+                            "Contract name mismatch: expected '{}', found '{}'",
+                            first_contract_name, root.contract_name
+                        ));
```
[the remark this answers is not in the document]
Comment (giovannidisiena) [reply src/hir/combiner.rs]: @alexfertel, I believe I have made some progress on this front, although I'm not quite sure what additional information we want to report and how. For example, I am having some difficulty working out how we can get the span of the original text from the HIR (more context on L125 of `combiner.rs`). Some input here would be great!

File: src/syntax/tokenizer.rs (hunk 11 above)
Comment (giovannidisiena) [reply src/syntax/tokenizer.rs]: @alexfertel, based on the latest commit, all tests pass. The issue above occurs when changing the `reset` function to set `identifier_mode` to true. The rationale here that, given `identifier_mode` is set to true by default since:
> The first token must always be a contract name, which has to be a valid Solidity identifier.

it should also be true here as we don't want this state to be changed to false when `reset` is called within `tokenize`; however, since the contract name is not sanitized before the tokenization step then the tests fail with `IdentifierCharInvalid('.')`.

Do you agree with this assessment? If so, how should we handle this?

File: src/hir/combiner.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (alexfertel) [reply src/hir/combiner.rs]: >  I am having some difficulty working out how we can get the span of the original text from the HIR

Ah, yes, this is a limitation of the current design, since a HIR doesn't really map to a tree, this is something to figure out. No worries, we don't have to tackle this here since it is unrelated and uphill.

File: src/hir/combiner.rs [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+                                    contract_definition = match &mut root.children[0] {
+                                        Hir::ContractDefinition(contract) => contract,
+                                        _ => unreachable!(),
+                                    }
+                                } else {
+                                    let text = contract_def.identifier.clone();
+                                    let identifier = get_contract_name_from_identifier(&text);
+                                    let accumulated_identifier = contract_definition.identifier.clone();
+                                    if identifier != accumulated_identifier {
+                                        let span = Span::default(); // @follow-up - how can we get the span from the HIR? Is it even necessary? This would be easier to do with verification of the AST. One option is to use the index of the HIR in the vector of HIRs since we know the identifier is the start of a given tree.
```
Comment (alexfertel) [inline src/hir/combiner.rs:125]: Yeah, this is a good idea, but I'm hesitant to commit until I think more about it. We can report with the empty span for now.

**Comment by alexfertel** [comment]
I don't have much time right now to review, but I will do so soon after New Year's!

**Comment by alexfertel** [comment]
Heyo @giovannidisiena ! Sorry about the delay, I've been very busy. Finally had some time to look at this.

Very good job! This is basically what I had in mind. What's left is testing, which I'll leave in your hands. I'd expect unit tests + integration tests in the `tests` folder.

I pushed a couple of commits: one with a few structural changes that are basically nits from me and another one fixing what you brought up -- you were right that `identifier_mode` should start as `false`. This was a bug given that we sanitize contract names now.

One more thing we need to add to the TODO list is that we need to handle trees that are more than `\n\n` apart. I think it's reasonable to just ignore any whitespace around and between the `\n\n`.

Thank you very much, excited about merging this PR!

File: src/syntax/tokenizer.rs (hunk 11 above)
Comment (alexfertel) [reply src/syntax/tokenizer.rs]: Yes, you were right; it's fine not to treat contract names as identifiers since we sanitize them later. Lmk if you find any bugs with this behavior.

**Comment by giovannidisiena** [comment]
Hey @alexfertel, that's not a problem at all. I'm also very busy for the foreseeable so will have to attempt to chip away at the testing as and when possible.

Brilliant, thank you for the assist there, and yes agreed regarding the additional whitespace.

> Thank you very much, excited about merging this PR!

Likewise!

**Comment by alexfertel** [comment]
Hey @giovannidisiena, I just wanted to check how we're feeling about this one. I'd really like this not to get stale, so if you're not up for it I'm happy to take over!

**Comment by giovannidisiena** [comment]
Absolutely @alexfertel, thanks for checking in. I may get some time toward the end of this week, so will keep you updated; otherwise, it may be best for you to take over since the next opportunity I have will likely be toward the end of March/April, which is obviously not ideal for keeping this from getting stale. I'll let you know how it goes!

File: src/utils.rs [hunk 17]
```diff
@@ -49,23 +49,21 @@ pub(crate) fn pluralize<'a>(count: usize, singular: &'a str, plural: &'a str) ->
...
-///
-/// This function is called during verification of the combined HIR.
-pub(crate) fn get_contract_name_from_identifier(identifier: &String) -> String {
-    let identifier_parts: Vec<&str> = identifier.split(CONTRACT_PART_SEPARATOR).collect();
-    return sanitize(identifier_parts[0]);
+pub(crate) fn get_contract_name_from_identifier(identifier: &str) -> String {
+    let contract_name = identifier
+        .split(CONTRACT_PART_SEPARATOR)
+        .next()
+        .expect("should not be empty");
```
Comment (giovannidisiena) [inline src/utils.rs:63]: @alexfertel if this expectation is violated, will this cause an unhandled panic? If so, does this need to be propagated to the calling function?

**Review by giovannidisiena — COMMENTED** [review]
Just looking for some clarification on this. Additionally, both `check` and `scaffold` are currently erroring with "a `Corner` must be the last child" – I'd thought this wouldn't be an issue since we are splitting the trees before forwarding on to the parser, but for some reason it is, and I haven't yet been able to track down the cause. If you have any ideas that would be great:)

File: src/utils.rs (hunk 17 above)
Comment (alexfertel) [reply src/utils.rs:63]: Ah, very good catch. That should actually be something like:

```rust
pub(crate) fn get_contract_name_from_identifier(identifier: &str) -> Option<String> {
    identifier
        .split(CONTRACT_PART_SEPARATOR)
        .next()
        .map(sanitize)
}
```

This is better because it reflects that the contract name might be malformed. Then at the call site, it would be:

```rust
                    let identifier = get_contract_name_from_identifier(&text).unwrap_or_default();
```
Since we handle the id being different a couple of lines below that one.

---

The problem is how do we handle the first tree having an empty tree. This leads me to think that we should fail on a previous compiler pass if we don't find the contract name.

So, to summarize, we should:
- Error if a tree starts with `::`, i.e. `::a_function`.
- Make `get_contract_name_from_identifier` panic if this assumption is not held, because that is a bug in `bulloak`

Does that make sense?

---

> Additionally, both check and scaffold are currently erroring with "a Corner must be the last child" – I'd thought this wouldn't be an issue since we are splitting the trees before forwarding on to the parser, but for some reason it is, and I haven't yet been able to track down the cause. If you have any ideas that would be great:)

Yeah, your mental model is correct, I'm unsure why this happens. I'll have a look tomorrow and see what that's about.

Comment (alexfertel) [reply src/utils.rs:63]: Update: I am currently unable to reproduce what you found.

I added a `multiple_roots.tree` file with the following contents:

```
Contract::function1
├── It should never revert.
└── When first arg is bigger than second arg
    └── It is all good

Contract::function2
└── When stuff happens
    └── It should do something simple
```

And it correctly generates the following contract:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.0;

contract Contract {
    function test_ShouldNeverRevert() external {
        // It should never revert.
    }

    function test_WhenFirstArgIsBiggerThanSecondArg() external {
        // It is all good
    }

    function test_WhenStuffHappens() external {
        // It should do something simple
    }
}
```

Maybe what's happening is that the last child on the tree you are testing with is not a `Corner` (`└`)?

Comment (giovannidisiena) [reply src/utils.rs:63]: Great, yeah that makes sense.

> Update: I am currently unable to reproduce what you found.

Weird. I'm still having this issue with the tree you provided, so quite probably a problem on my end. Having said that, I haven't really changed anything, so if hard reset doesn't work, then I'm stumped. `cargo build --release` should build with the latest changes, right?

Comment (alexfertel) [reply src/utils.rs:63]: Hmmm, you can clean your local setup with `cargo clean`, and then what happens if you do `cargo run -- scaffold tests/scaffold/multiple_roots.tree`?

Also, maybe you have commits that haven't been pushed?

**Review by alexfertel — COMMENTED** [review]
This is looking awesome! I want to do some checks myself locally, which I'll do as soon as I get some time, hopefully today. Thanks for all this work ❤️

File: tests/check/missing_contract_identifier.tree [hunk 18]
```diff
@@ -0,0 +1,5 @@
+ContractName
+└── It should find a contract identifier.
+
+::orphanedFunciton
```
[the remark this answers is not in the document]
Comment (giovannidisiena) [reply tests/check/missing_contract_identifier.tree:4]: Every single time 😂

**Comment by giovannidisiena** [comment]
> This is looking awesome! I want to do some checks myself locally, which I'll do as soon as I get some time, hopefully today. Thanks for all this work ❤️

Great! And thanks for all the help ❤️

**Comment by alexfertel** [comment]
Okay, local checks done!

> any violation cases to be fixed when running bulloak check --fix
Nothing comes to mind right now, but any work there can be done in a different PR, it doesn't need to be a part of this one. If you think of anything feel free to open an issue.

I think what's missing are the `Combiner` unit tests, and then I think we need to check the README and update it to reflect the new functionality. Does that make sense to you?

**Comment by giovannidisiena** [comment]
> Nothing comes to mind right now, but any work there can be done in a different PR, it doesn't need to be a part of this one. If you think of anything feel free to open an issue.

The only thing I can think of at the moment is perhaps fixing the contract name mismatch by applying the first identifier to all subsequent roots, but I'm happy to open it as a separate issue if you think it's not worth tackling right now.

**Comment by giovannidisiena** [comment]
@alexfertel, `Combiner` tests have been added, and the corresponding logic has been updated accordingly. I'm not sure I'll get around to refactoring the structural match violation logic, although if you are willing to provide some pointers, then that could speed things up, and I may be able to take a crack at it.

File: src/hir/combiner.rs [hunk 19]
```diff
@@ -157,3 +170,162 @@ impl Combiner {
...
+        let result = catch_unwind(|| combine(&hirs));
+        assert!(result.is_err());
+        assert_eq!(
+            result
+                .unwrap_err()
+                .downcast_ref::<String>()
+                .unwrap()
+                .as_str(),
+            "expected contract identifier at tree root"
+        );
```
Comment (alexfertel) [inline src/hir/combiner.rs:250]: Oh, this is neat. However, you can just annotate the test with `#[should_panic]`.
This is a tiny bit more precise so I'm ok with it 👍🏻

**Review by alexfertel — APPROVED** [review]
LGTM! 🚀

File: src/hir/combiner.rs (hunk 19 above)
Comment (giovannidisiena) [reply src/hir/combiner.rs:250]: Thanks for the heads up 🙏

**Comment by alexfertel** [comment]
> I'm not sure I'll get around to refactoring the structural match violation logic

That's okay, that part of the code is not polished and needs lots of love. I'm merging this in!

When you're happy with this, feel free to change it to ready for review.

**Comment by giovannidisiena** [comment]
@alexfertel just realised I haven't updated the README, doing that now.

**Comment by alexfertel** [comment]
I forgot about that as well, thank you!

**Comment by giovannidisiena** [comment]
@alexfertel While writing up an example for the README, I noticed that the resulting Solidity test function names did not include the root identifier function name, so I have also added that and refactored the `Combiner`. May be worth another review to make sure you're still happy with it.

**Review by alexfertel — APPROVED** [review]
LGTM! 🚀
